src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -350,20 +337,26 @@ async def mypy_typecheck_partition(
 async def mypy_determine_partitions(
     request: MyPyRequest, mypy: MyPy, python_setup: PythonSetup
 ) -> MyPyPartitions:
-
-    resolve_and_interpreter_constraints_to_coarsened_targets = (
-        await partition._by_interpreter_constraints_and_resolve(request.field_sets, python_setup)
+    resolve_and_interpreter_constraints_to_field_sets = (
+        _partition_by_interpreter_constraints_and_resolve(request.field_sets, python_setup)
+    )
+    coarsened_targets = await Get(
+        CoarsenedTargets,
+        CoarsenedTargetsRequest(field_set.address for field_set in request.field_sets),
     )
+    coarsened_targets_by_address = coarsened_targets.by_address()
 
     return MyPyPartitions(
         MyPyPartition(
-            FrozenOrderedSet(roots),
-            FrozenOrderedSet(CoarsenedTargets(root_cts).closure()),
+            FrozenOrderedSet(field_sets),
+            CoarsenedTargets(
+                coarsened_targets_by_address[field_set.address] for field_set in field_sets
+            ),
```

Comment:
The `CoarsenedTargets` constructor doesn't dedup, so this may get you the same CT multiple times. Mostly harmless though, I think.

Reply:
I `OrderedSet`ted it